Notify the congestion controller of losses first (#3540)
quicwg/base-drafts: draft-ietf-quic-recovery.md

    largest_acked_packet[pn_space] =
        max(largest_acked_packet[pn_space], ack.largest_acked)

>   // Determine which packets have been acknowledged and remove

Reviewer: ```suggestion
  // DetectNewlyAckedPackets finds packets have been newly acknowledged and removes
```

  // them from sent_packets.

>   newly_acked_packets = DetectNewlyAckedPackets(ack, pn_space)

Reviewer: ```suggestion
  newly_acked_packets = DetectAndRemoveAckedPackets(ack, pn_space)
```
here and below

    largest_acked_packet[pn_space] =
        max(largest_acked_packet[pn_space], ack.largest_acked)

>   // DetectNewlyAckedPackets finds packets that are newly

Reviewer 2: `DetectNewlyAckedPackets` is not defined in `master`.

  if (ACK frame contains ECN information):
      ProcessECN(ack, pn_space)

>   lost_packets = DetectLostPackets(pn_space)

Reviewer: ```suggestion
  lost_packets = DetectAndRemoveLostPackets(pn_space)
```
here and below

    GetEarliestTimeAndSpace(loss_time)
  if (earliest_loss_time != 0):
    // Time threshold loss Detection
    lost_packets = DetectLostPackets(pn_space)

>     OnPacketsLost(lost_packets)

Reviewer: ```suggestion
    if (!lost_packets.empty()):
      OnPacketsLost(lost_packets)
```

Author: If a timeout occurred, then lost_packets should never be empty I believe?  I could add an assert for that if you'd like?

Reviewer: Hm, that's true. I was thinking about stream cancellations, but that doesn't matter here. assert SGTM

## On Packet Acknowledgement

> Invoked from loss detection's OnPacketAcked and is supplied with the

Reviewer: ```suggestion
Invoked from loss detection's OnAckReceived and is supplied with the
```

newly acked_packets from sent_packets.

~~~

>    InLossRecovery(sent_time):

Reviewer: ```suggestion
   InCongestionRecovery(sent_time):
```
Revert this change. Was this deliberate? The term congestion recovery is used earlier in the spec and the term is defined.

Author: This was deliberate.  I've always heard it called loss recovery or just 'the recovery period'

But I can revert it.

     return sent_time <= congestion_recovery_start_time

   OnPacketsAcked(acked_packets):
     for (packet in acked_packets):
       // Remove from bytes_in_flight.
       bytes_in_flight -= packet.size

>        if (InLossRecovery(packet.time_sent)):

Reviewer: ```suggestion
       if (InCongestionRecovery(packet.time_sent)):
```

Invoked from DetectLostPackets when packets are deemed lost.

~~~
   InPersistentCongestion(lost_packets):
     pto = smoothed_rtt + max(4 * rttvar, kGranularity) +
       max_ack_delay
     congestion_period = pto * kPersistentCongestionThreshold
     // Determine if all packets in the time period before the

>      // newest lost packet, including the edges, are marked

Reviewer: ```suggestion
     // largest newly lost packet, including the edges, are marked
```

Author: I'm not sure largest makes this less clearer, at least to me.

>      CongestionEvent(largest_lost_packet.time_sent)

Reviewer: How about also changing this to the following, and removing the line before this (last() is a language-specific method, so I'm suggesting largest())?
```suggestion
     CongestionEvent(lost_packets.largest().time_sent)
```

Author: I'm not a big fan of largest here and above.  If you want to make the change, maybe another PR, since this one is already quite large.
